### PR TITLE
feat(secrets): improve detect secrets checks and add config

### LINKED ISSUE #3905: Add configuration for secret related checks
### New feature motivation

All checks related to secret detection lead to a massive amount of false positives. This could get prevented if there would be a possibility to configure those checks, since most of them use the `detect-secrets`  package. 

Use case: 
The checks [awslambda_function_no_secrets_in_variables](https://github.com/prowler-cloud/prowler/blob/484cf6f49d558da344b2f8dd149d831b693c9c2d/prowler/providers/aws/services/awslambda/awslambda_function_no_secrets_in_variables/awslambda_function_no_secrets_in_variables.py#L12) and [ecs_task_definitions_no_environment_secrets](https://github.com/prowler-cloud/prowler/blob/484cf6f49d558da344b2f8dd149d831b693c9c2d/prowler/providers/aws/services/ecs/ecs_task_definitions_no_environment_secrets/ecs_task_definitions_no_environment_secrets.py#L12) do use the [detect-secrets](https://github.com/Yelp/detect-secrets) package.

Unfortunately there is no possibility (at least no documented one) that allows to configure the checks done with that package. The repository documents how you can [use a customized settings file](https://github.com/Yelp/detect-secrets?tab=readme-ov-file#usage-in-other-python-scripts), which can for example adjust the threshold of `Base64HighEntropyString` etc. 

Could it be possible to add the JSON like config to the prowler's [config.yaml](https://github.com/prowler-cloud/prowler/blob/master/prowler/config/config.yaml)? 

Furthermore it would be nice to have an option to filter out environment variable names. 

Example: 
In an AWS `TaskDefinition` i use a variable named `GCP_SECRET_ARN`. This contains an ARN to `AWS_SECRETS_MANAGER` service which will be used during runtime of a script to retrieve a JSON, since it is not really an option to mount a JSON with newline characters into an environment variable as secret. Thus i would like to filter out for name patterns like "`*SECRET_ARN`". 
Would this make sense to you, too? 

As documented for `detect-secrets`, one could use the following parameter:
```
  --exclude-secrets EXCLUDE_SECRETS
                        If secrets match this regex, it will be ignored.
```

This could already help me out for my use-case. 

### Solution Proposed

Add an option to the config.yaml that would apply to all scans which use `detect-secrets`. 

My pseudo example: 

```yaml
aws:
  # configure entropy threshold
  awslambda_function_no_secrets_in_variables_Base64HighEntropyString_limit: 7.0
  # Configure a  regex-based ignore-pattern for variable names
  ecs_task_definitions_no_environment_secrets_ignore_patterns:
    - ".*SECRET_ARN"
    - ".*PASSWORD_ARN"
  ```

And then fetch those values in the corresponding python scripts... 

I know: configuring this is more a global option than a "per-check" option. So maybe it makes even more sense to create a default config-file for `detect-secrets` which someone could adjust. Just as with the yaml files. 

### Describe alternatives you've considered

I did consider to overwrite the python scripts for the mentioned checks, but this is somewhat unfeasible to have it within a deployment CI-CD deployment.. It would be more easy if i could just specify or change a global settings file for the `detect-secrets` package. 

### Additional context

Maybe i have overseen something in the documentation or something in the code itself. So if there already is a solution for my presented use cases, i would already like to apologize.

### LINKED ISSUE #4893: Improve `detect-secrets` engine: configuration + plugins
### Steps to Reproduce

Prowler relies on https://github.com/Yelp/detect-secrets to detect secrets for some AWS resources in some checks:

- autoscaling_find_secrets_ec2_launch_configuration
- awslambda_function_no_secrets_in_code
- awslambda_function_no_secrets_in_variables
- cloudformation_stack_outputs_find_secrets
- cloudwatch_log_group_no_secrets_in_logs
- ec2_instance_secrets_user_data
- ec2_launch_template_no_secrets
- ecs_task_definitions_no_environment_secrets
- ssm_document_secrets

Currently the detection engine is raising some false positives we need to work on to improve detection.

Some examples:
- `MYSQL_ALLOW_EMPTY_PASSWORD` is flagged as a secret when the value is just `yes` or `no`.
- `detect-secrets` TelegramBotToken plugin is flagging some strings to be a secret when it is not.

### Expected behavior

Prowler should not raise false positives regarding secrets detection.

- We need to investigate more about adding/removing plugins.
- Adjust the `base64-limit` and the `hex-limit`

### Actual Result with Screenshots or Logs

N/A

### How did you install Prowler?

From pip package (pip install prowler)

### Environment Resource

N/A

### OS used

N/A

### Prowler version

4.3.5

### Pip version

24.0

### Context

_No response_

### PR DESCRIPTION
**Pull request by pedrooot**
### Context

Fix #4893
Fix #3905

### Description

Add `secrets` configuration to make the output more accurate. Checks modified:
- `autoscaling_find_secrets_ec2_launch_configuration `
- `awslambda_function_no_secrets_in_code `
-  `awslambda_function_no_secrets_in_variables`
-  `cloudformation_stack_outputs_find_secrets`
-  `cloudwatch_log_group_no_secrets_in_logs`
-  `ec2_instance_secrets_user_data`
-  `ec2_launch_template_no_secrets`
-  `ecs_task_definitions_no_environment_secrets`
-  `ssm_document_secrets`
- `codebuild_project_no_secrets_in_variables`

Also, `config.yaml` will have the field `secrets_ignore_patterns`, a list that can be changed to set a list of regex from your known secrets. This way, `detect-secrets` will discard the secrets that match the regex.

### Checklist

- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/lib/utils/utils.py [hunk 1]
```diff
@@ -80,31 +81,57 @@ def hash_sha512(string: str) -> str:
...
-    temp_data_file = tempfile.NamedTemporaryFile(delete=False)
-    temp_data_file.write(bytes(data, encoding="raw_unicode_escape"))
-    temp_data_file.close()
+def is_known_non_secret(line):
+    known_non_secrets = [
+        "MYSQL_ALLOW_EMPTY_PASSWORD",
+        "GCP_SECRET_ARN",
+        "S3_BUCKET",
+        "AWS_STACK",
+    ]
```
Comment (jfagoagas) [inline prowler/lib/utils/utils.py]: I think this list could end up with thousands of values, I'm not sure about placing it here. Do we have any other way not to flag them as secrets?

File: prowler/lib/utils/utils.py [hunk 2]
```diff
@@ -80,31 +81,57 @@ def hash_sha512(string: str) -> str:
...
 
     secrets = SecretsCollection()
     with transient_settings(
         {
             "plugins_used": [
-                {"name": "Base64HighEntropyString", "limit": 4.5},
-                {"name": "HexHighEntropyString", "limit": 3.0},
+                {"name": "KeywordDetector"},
+                {"name": "Base64HighEntropyString", "limit": 4.0},
+                {"name": "HexHighEntropyString", "limit": 1.5},
```
Comment (jfagoagas) [inline prowler/lib/utils/utils.py]: Why did you chose these values? Can you document it?

File: prowler/lib/utils/utils.py [hunk 3]
```diff
@@ -80,31 +81,57 @@ def hash_sha512(string: str) -> str:
...
                 {"name": "AWSKeyDetector"},
             ],
             "filters_used": [
                 {"path": "detect_secrets.filters.common.is_invalid_file"},
-                {"path": "detect_secrets.filters.heuristic.is_likely_id_string"},
+                {"path": "detect_secrets.filters.common.is_known_non_secret"},
+                {"path": "detect_secrets.filters.common.is_known_false_positive"},
```
Comment (jfagoagas) [inline prowler/lib/utils/utils.py]: What is the purpose of these filters?

File: tests/providers/aws/services/awslambda/awslambda_function_no_secrets_in_code/awslambda_function_no_secrets_in_code_test.py [hunk 4]
```diff
@@ -107,7 +107,7 @@ def test_function_code_with_secrets(self):
             assert result[0].status == "FAIL"
             assert (
                 result[0].status_extended
-                == f"Potential secret found in Lambda function {LAMBDA_FUNCTION_NAME} code -> lambda_function.py: Secret Keyword on line 3."
+                == f"Potential secrets found in Lambda function {LAMBDA_FUNCTION_NAME} code -> lambda_function.py: Base64 High Entropy String on line 3, Secret Keyword on line 3; lambda_function.py: Base64 High Entropy String on line 3, Secret Keyword on line 3."
```
Comment (jfagoagas) [inline tests/providers/aws/services/awslambda/awslambda_function_no_secrets_in_code/awslambda_function_no_secrets_in_code_test.py]: Why this change? We should not flag for the same secret several times.

File: tests/providers/aws/services/awslambda/awslambda_function_no_secrets_in_variables/awslambda_function_no_secrets_in_variables_test.py [hunk 5]
```diff
@@ -115,6 +115,49 @@ def test_function_secrets_in_variables(self):
...
+        function_arn = f"arn:aws:lambda:{AWS_REGION_US_EAST_1}:{AWS_ACCOUNT_NUMBER}:function/{function_name}"
+
+        lambda_client.functions = {
+            "function_name": Function(
+                name=function_name,
+                security_groups=[],
+                arn=function_arn,
+                region=AWS_REGION_US_EAST_1,
+                runtime=function_runtime,
+                environment={"TELEGRAM_BOT_TOKEN": "telegram-token"},
```
Comment (jfagoagas) [inline tests/providers/aws/services/awslambda/awslambda_function_no_secrets_in_variables/awslambda_function_no_secrets_in_variables_test.py:131]: You should add a valid Telegram Bot Token because detect-secrets flags it as a secret for the environment value not by their name.

File: tests/providers/aws/services/ec2/ec2_instance_secrets_user_data/ec2_instance_secrets_user_data_test.py [hunk 6]
```diff
@@ -256,7 +300,7 @@ def test_one_ec2_file_with_secrets_gzip(self):
             assert result[0].status == "FAIL"
             assert (
                 result[0].status_extended
-                == f"Potential secret found in EC2 instance {instance.id} User Data -> Secret Keyword on line 1, Hex High Entropy String on line 3, Secret Keyword on line 3, Secret Keyword on line 4."
+                == f"Potential secret found in EC2 instance {instance.id} User Data -> Secret Keyword on line 1, Hex High Entropy String on line 3, Secret Keyword on line 3, Hex High Entropy String on line 4, Secret Keyword on line 4."
```
Comment (jfagoagas) [inline tests/providers/aws/services/ec2/ec2_instance_secrets_user_data/ec2_instance_secrets_user_data_test.py]: The same as before, why are we adding more for the same secret?

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: prowler/lib/utils/utils.py (hunk 1 above)
Comment (jfagoagas) [reply prowler/lib/utils/utils.py]: I think I'll remove that for now until we explore all the options to try not to flag them as secrets.

Comment (jfagoagas) [reply prowler/lib/utils/utils.py]: Let's try to improve the configuration for the test that will fail once we remove this block.

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: prowler/lib/utils/utils.py [hunk 7]
```diff
@@ -81,29 +80,26 @@ def hash_sha512(string: str) -> str:
...
-    common.is_known_non_secret = is_known_non_secret
-    if not file:
-        temp_data_file = tempfile.NamedTemporaryFile(delete=False)
-        temp_data_file.write(bytes(data, encoding="raw_unicode_escape"))
-        temp_data_file.close()
-
-    secrets = SecretsCollection()
-    with transient_settings(
-        {
+def detect_secrets_scan(data=None, file=None, excluded_secrets: list[str] = None):
```
Comment (jfagoagas) [inline prowler/lib/utils/utils.py]: Add the types for the arguments and the return value please.

File: tests/providers/aws/services/ec2/ec2_instance_secrets_user_data/ec2_instance_secrets_user_data_test.py (hunk 6 above)
Comment (jfagoagas) [reply tests/providers/aws/services/ec2/ec2_instance_secrets_user_data/ec2_instance_secrets_user_data_test.py]: Review this please.

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: prowler/lib/utils/utils.py (hunk 7 above)
Comment (pedrooot) [reply prowler/lib/utils/utils.py]: Done

**Review by jfagoagas — DISMISSED** [review]
What a great improvement in our way to detect-secrets!

**Review by jfagoagas — APPROVED** [review]

**Comment by cetteup** [comment]
@pedrooot This is a great improvement! When will this be released?
